### PR TITLE
WIP: remove zz_generated.*.go files from the repo

### PR DESCRIPTION
**Pull request by fejta**
Please see https://github.com/kubernetes/kubernetes/pull/62823

Originally
<details>
* Create a `k8s_deepcopy_all()` `genrule` which generates `zz_generated.deepcopy.go` files for any package that requests it with a `// +k8s:deepcopy=` comment
* Create a `k8s_deepcopy()` macro for a `genrule` which extracts `zz_generated.deepcopy.go` for a specific package
* update `hack/update-bazel.sh` to automatically manage these two rules.

Remaining work for this PR:
* Ensure this doesn't cause problems for the publishing bot
* Ensure this runs on both linux and mac
* General refactoring for clarify/sanity/etc

Work for future PRs:
* Investigate [auto-generating](https://github.com/fejta/test-infra/tree/autogo/autogo) these rules without `hack/update-bazel.sh` 
* Refactor the rule into something which will handle any of the `zz_generated.FOO.go` files
</details>

/hold

Appreciate any comments feedback

### CODE REVIEW

File: build/deepcopy.bzl [hunk 1]
```diff
@@ -0,0 +1,52 @@
+_cmd = """
+  # location of prebuilt deepcopy generator
+  dcg=$$PWD/'$(location //vendor/k8s.io/code-generator/cmd/deepcopy-gen)'
+  # original pwd
+  export O=$$PWD
+  # gopath/goroot for genrule
+  export GOPATH=$$PWD/.go
```
Comment (ixdy) [inline build/deepcopy.bzl]: the `go_genrule` rule from repo-infra probably does what you want here.

Comment (ixdy) [reply build/deepcopy.bzl]: I'm not sure what it's doing incorrectly with staging directories, but probably.

The other alternative is using the `go_path` rule from `rules_go`, which does similar things to what you're trying to do here. I had it mostly working with our existing bazel code generators (as an alternative to `go_genrule`) though there were still a few lingering issues. (I'm also not sure where that commit snuck off to.)

Comment (fejta) [reply build/deepcopy.bzl]: We need to ignore `$GOPATH/src/k8s.io/kubernetes/staging/src/FOO/` and instead look in `$GOPATH/src/k8s.io/kubernetes/vendor/FOO` instead

Comment (ixdy) [reply build/deepcopy.bzl]: `go_genrule` just creates a new go workspace and symlinks to the original tree, similar to what `hack/run-in-gopath.sh` does (and which is what make does, I think).

does deepcopy not read through the vendor symlinks?

File: build/deepcopy.bzl [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  export GOROOT=/usr/lib/google-golang
+
+  # symlink in source into new gopath
+  mkdir -p $$GOPATH/src/k8s.io
+  ln -snf $$PWD $$GOPATH/src/k8s.io/kubernetes
+  # symlink in all the staging dirs
+  for i in $$(ls staging/src/k8s.io); do
+    ln -snf $$PWD/staging/src/k8s.io/$$i $$GOPATH/src/k8s.io/$$i
```
Comment (ixdy) [inline build/deepcopy.bzl]: or do you mean that all of the staging dirs need to be installed in `$GOPATH/src/k8s.io/`, rather than in vendor?

**Comment by sttts** [comment]
/hold

These kind of changes will break all staging repos, i.e. client-go and friends. We have to recreate the generated files for them. I could imagine to have a `hack/update-codegen.sh` in each of the staging repos doing this, and calling that from the publishing bot. But this needs quite some work in the bot.

**Comment by sttts** [comment]
This will also break the vendoring work flow of all 3rdparty consumers of the k/k repo.

**Comment by fejta** [comment]
Fear not, this is WIP for a reason and not in danger of getting checked in.

This PR presumes that the longstanding project goal of removing checked in code from k/k remains is the right goal -- https://github.com/kubernetes/kubernetes/blob/master/.gitignore#L111 -- primarily due to the rebase pain it causes.

Would you mind if we move the conversation about whether we want to check in generated code to somewhere other than this PR? Relevant issue: https://github.com/kubernetes/kubernetes/issues/8830

Totally open to the idea that checking in generated code is the right thing to do, and am not planning to merge this before that is resolved.

File: build/deepcopy.bzl (hunk 2 above)
Comment (fejta) [reply build/deepcopy.bzl]: Yes, although TBH I'm not really sure if this is necessary.

Between our staging pretend repo, symlinks galore and golang's vendoring logic I have no idea what it is supposed to look like. Solution: create symlinks until it works and distill later 😈

File: build/deepcopy.bzl (hunk 1 above)
Comment (fejta) [reply build/deepcopy.bzl]: Thanks for the tip!

Refactored to use `go_genrule`

**Comment by fejta** [comment]
http://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/62485/pull-kubernetes-verify/86390/

**Comment by lavalamp** [comment]
Awesome. Last time we tried something like this we got pushback because the repo was no longer buildable via ordinary means (go get). Have the people doing that given up yet? We should publicize this.

**Comment by fejta** [comment]
The info I'm getting from Tim is that we definitely want to do this but bazel is blocking us from making progress. 

This is to ensure that bazel isn't blocking us.

**Comment by lavalamp** [comment]
Yes, that's the last thing I knew on this, too.

**Comment by ixdy** [comment]
There's already a few things we don't check in, notably the openapi-generated files and (one of) the go-bindata files (for e2e tests).

File: hack/update-bazel.sh [hunk 3]
```diff
@@ -35,18 +39,132 @@ go install ./vendor/github.com/bazelbuild/bazel-gazelle/cmd/gazelle
...
+echo Running gazelle to cleanup any changes...
+gazelle-fix
+
+if ! which -s bazel; then
+  return 0  # We should have already generated these files
+fi
+
+find . -name BUILD -or -name BUILD.bazel | xargs sed -i '' -e 's|@io_bazel_rules_go//go:def.bzl|//build:go.bzl|g'
+
+# Add dependencies from generated files
```
Comment (ixdy) [inline hack/update-bazel.sh]: oh wow, the kubernetes codegen is gross and rather bazel-hostile. :\

Comment (fejta) [reply hack/update-bazel.sh]: Autogo might be able to help here

**Comment by thockin** [comment]
This is an enormous PR and I am ill-equipped to review it.  How important is it that literal `bazel build` be the human interface vs `make bazel-build` or similar ?  I am wary of replicating this logic, now that I see it laid out.  For example, I have https://github.com/kubernetes/kubernetes/pull/54372 - how much of that would need to be done twice?

File: hack/update-bazel.sh [hunk 4]
```diff
@@ -17,6 +17,10 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+if ! which -s bazel && [[ "${CALLED_FROM_MAIN_MAKEFILE:-}" == "" ]]; then
```
Comment (thockin) [inline hack/update-bazel.sh]: explain this?  It is not true for most hack scripts.

File: hack/update-bazel.sh [hunk 5]
```diff
@@ -35,18 +39,132 @@ go install ./vendor/github.com/bazelbuild/bazel-gazelle/cmd/gazelle
 
 go install ./vendor/github.com/kubernetes/repo-infra/kazel
 
+cd "${KUBE_ROOT}"
+go get github.com/bazelbuild/buildtools/buildozer
```
Comment (thockin) [inline hack/update-bazel.sh]: would need to vendor buildozer, same as gazelle and kazel

File: hack/update-bazel.sh [hunk 6]
```diff
@@ -35,18 +39,132 @@ go install ./vendor/github.com/bazelbuild/bazel-gazelle/cmd/gazelle
...
-    -external=vendored \
-    -proto=legacy \
-    -mode=fix
-# gazelle gets confused by our staging/ directory, prepending an extra
-# "k8s.io/kubernetes/staging/src" to the import path.
-# gazelle won't follow the symlinks in vendor/, so we can't just exclude
-# staging/. Instead we just fix the bad paths with sed.
-find staging -name BUILD -o -name BUILD.bazel | \
-  xargs ${SED} -i 's|\(importpath = "\)k8s.io/kubernetes/staging/src/\(.*\)|\1\2|'
+dozer() {
```
Comment (thockin) [inline hack/update-bazel.sh]: comments?  I have no clue what this is doing

File: staging/src/k8s.io/client-go/examples/workqueue/BUILD [hunk 7]
```diff
@@ -1,7 +1,7 @@
 package(default_visibility = ["//visibility:public"])
 
 load(
-    "@io_bazel_rules_go//go:def.bzl",
+    "//build:go.bzl",
```
Comment (thockin) [inline staging/src/k8s.io/client-go/examples/workqueue/BUILD]: Can these changes get a separate PR, just to make this easier to scan - thousands of files, many with justthis change

Comment (fejta) [reply staging/src/k8s.io/client-go/examples/workqueue/BUILD]: Check out https://github.com/kubernetes/kubernetes/pull/62823

File: hack/update-bazel.sh (hunk 4 above)
Comment (fejta) [reply hack/update-bazel.sh]: If we are going to delete the generated code we need to generate the files in order to calculate them. 

Thus the user either needs to have bazel installed (which will generate the files) or use the Makefile rule to run updates (which adds a dep on the generate rule)

**Comment by fejta** [comment]
@thockin thanks for the comments. Please see https://github.com/kubernetes/kubernetes/pull/62823 which breaks down this PR into the core bazel rules.

I'm opposed to putting `make` in front of `bazel` -- people who want to use make can already use make.

 #54372 is fantastic and we could just change the genrule to parse the output from that program rather than `go list`

**Comment by fejta** [comment]
PS feel free to assign me your PR if you want it in and think that will help

**Comment by fejta** [comment]
@ixdy this PR still useful to you at all?

**Comment by ixdy** [comment]
We can close. I think there's still no verdict on whether we want to remove the generated files from git.

**Pull request closed without merging**